src/test/kotlin/com/buildit/bookit/auth/JwtAuthenticationFilterUnitTests.kt

```diff
@@ -0,0 +1,103 @@
+package com.buildit.bookit.auth
+
+import com.nhaarman.mockito_kotlin.doReturn
+import com.nhaarman.mockito_kotlin.mock
+import com.nhaarman.mockito_kotlin.never
+import com.nhaarman.mockito_kotlin.times
+import com.nhaarman.mockito_kotlin.verify
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Nested
+import org.junit.jupiter.api.Test
+import org.springframework.mock.web.MockHttpServletRequest
+import org.springframework.mock.web.MockHttpServletResponse
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
+import org.springframework.security.core.context.SecurityContext
+import javax.servlet.FilterChain
+
+class JwtAuthenticationFilterUnitTests {
+    private val goodSpringAuthToken = UsernamePasswordAuthenticationToken("principal", "credentials")
+    private val nullSpringAuthToken: UsernamePasswordAuthenticationToken? = null
+
+    private val filterChain: FilterChain = mock()
+
+    private val securityContext: SecurityContext = mock()
+
+    private lateinit var request: MockHttpServletRequest
+    private lateinit var response: MockHttpServletResponse
+
+    private lateinit var securityContextHolderWrapper: SecurityContextHolderWrapper
+
+    @BeforeEach
+    fun `set up`() {
+        request = MockHttpServletRequest()
+        response = MockHttpServletResponse()
+
+        securityContextHolderWrapper = mock {
+            on { obtainContext() }.doReturn(securityContext)
+        }
+    }
+
+    @Nested
+    inner class `with well-formed header` {
+        @Test
+        fun `successful verification`() {
+            request.addHeader("Authorization", "Bearer test")
+
+            val jwtAuthenticator = mock<JwtAuthenticator> {
+                on { getAuthentication("test", request) }
+                    .doReturn(goodSpringAuthToken)
+            }
+
+            val filter = JwtAuthenticationFilter(mock {}, jwtAuthenticator, securityContextHolderWrapper)
+            filter.doFilter(request, response, filterChain)
+            verify(securityContext, times(1)).authentication = goodSpringAuthToken
```

Comment:
this seems... weird...what's this really testing?  Seems there should be assert somewhere...

Reply:
The verify() is an “assertion” that the expected Spring Auth Token is assigned once (which is the side-effect and purpose of the filter).  Same goes for all the verify() calls in the test.  Let me know if I’m missing something.

Reply:
I guess I’m seeing this piece of code out of context, so I’m not sure what is confusing about it, esp. if you accept verify() as a technique in the next block, below.

Reply:
Just that the `.authentication = goodSpringAuthToken` is outside the verify()... rather than mocking the securitycontextholder, just use the real one and assert the authentication that was set == goodSpringAuthToken?

Reply:
could also be that it's been a while since i've seen verify's

Reply:
Yeah, it looks weird, but it's not that different than what's being done with `.doFilter(request, response)` below.  If Kotlin would allow the setter to look normal (`.setAuthentication(goodSpringAuthToken)`), it wouldn't look as weird (?).

Reply:
maybe...well it does what it says... sounds good to me